### PR TITLE
Go: extract and expose struct tags, interface method IDs

### PR DESCRIPTION
**Pull request by smowton**
This enables us to distinguish all database types in QL. Previously structs with the same field names and types but differing tags, and interface types with matching method names and at least one non-exported method but declared in differing packages, were impossible or only sometimes possible to distinguish in QL. With this change these types can be 
distinguished, as well as permitting queries to examine struct field tags, e.g. to read JSON field name associations.

This is a pre-requisite to (some approaches to) dealing with Go 1.23's more direct exposure of type aliases, since it enables us to distinguish all types that are distinct in the database in QL, and therefore implement up-to-aliasing type matching, known in the Go spec as _identical types_.

### CODE REVIEW

File: go/ql/lib/change-notes/2024-09-03-tags-and-interface-ids.md
Comment (owen-mc) [inline go/ql/lib/change-notes/2024-09-03-tags-and-interface-ids.md:5]:
```suggestion
* Added member predicates `StructTag.hasOwnFieldWithTag` and `Field.getTag`, which enable CodeQL queries to examine struct field tags.
* Added member predicate `InterfaceType.getMethodTypeById`, which enables CodeQL queries to distinguish interfaces with matching non-exported method names that are declared in different packages, and are therefore incompatible.
```

File: go/extractor/dbscheme/tables.go
Comment (owen-mc) [inline go/extractor/dbscheme/tables.go:1155]:
```suggestion
	EntityColumn(StructType, "parent"),
```
Tags only exist on fields of structs, so I don't see why we should make this table more general than that. (Various names should change as well, of course.)

**Review by owen-mc — COMMENTED** [review]
Longer review to follow.

File: go/extractor/extractor.go
Comment (mbg) [inline go/extractor/extractor.go:1567]: Question: My understanding is that, in Go, exported methods always start with an upper-case character. Why did you go for this `meth.Id() != meth.Name()` check rather than checking the first character of the name? Did that not work or do you think this method is more reliable or has other advantages?

File: go/extractor/dbscheme/tables.go
Comment (mbg) [inline go/extractor/dbscheme/tables.go:1160]: If I understand correctly, this table includes entries for methods (by index in the interface) which are private. If so, I don't think the comment makes that very clear. The current comment suggests that every _interface type_ has _a private method id_. How about:
```suggestion
// InterfacePrivateMethodIdsTable is the table associating interface types with the indices of their private methods.
```

File: go/ql/lib/semmle/go/Types.qll
Comment (mbg) [inline go/ql/lib/semmle/go/Types.qll:770]: This is probably more a complaint about Go's terminology than anything else, but `id` doesn't seem like the best choice of name for this.

File: go/ql/lib/semmle/go/Types.qll
Comment (mbg) [inline go/ql/lib/semmle/go/Types.qll:768]: It may be good to extend the example to include the `id`s of the methods (with sample packages) to show explicitly what you mean in the previous paragraph.

File: go/ql/test/library-tests/semmle/go/Types/InterfaceIds.ql
Comment (mbg) [inline go/ql/test/library-tests/semmle/go/Types/InterfaceIds.ql:1]: Should this file be called `InterfaceMethodIds.ql`?

**Review by mbg — COMMENTED** [review]
Broadly looks good! Thank you for improving this and moving it out into it's own PR. I just have a few suggestions in addition to @owen-mc's comments, which also make sense.

Also, to sanity check: in the PR description you discuss that part of the motivation here is to be able to distinguish types better. That makes sense and I found the relevant part of the Go specification for this in https://go.dev/ref/spec#Type_identity. For structs:

> Two struct types are identical if they have the same sequence of fields, and if corresponding fields have the same names, and identical types, and identical tags. [Non-exported](https://go.dev/ref/spec#Exported_identifiers) field names from different packages are always different.

For interfaces:

> Two interface types are identical if they define the same type set.

Looking over the tests here, I can see that the tests exercise the new functionality and that seems to behave as expected. Do the tests cover the new ability to decide (in)equality that you are hoping for? Could you comment on how the tests cover that?

**Comment by owen-mc** [comment]
Tests failing:
```
 ./ql/test/library-tests/semmle/go/Types/pkg2/tst.go
Error: make: *** [Makefile:15: check-formatting] Error 1
./ql/test/library-tests/semmle/go/Types/struct_tags.go
Error: Process completed with exit code 2.
```

**Comment by smowton** [comment]
Retargeted this against `main` because we're currently not expecting to need this for rc/3.15 if we go for a simpler alias-erasing approach in the interim

File: go/extractor/dbscheme/tables.go
Comment (owen-mc) [reply go/extractor/dbscheme/tables.go:1160]: I agree it should be clearer. This is slightly more informative:
```suggestion
// InterfacePrivateMethodIdsTable is the table associating interface types with the indices and ids of their private methods.
```

**Review by owen-mc — COMMENTED** [review]
Good work spotting these problems and fixing them. A few small suggestions for improvement.

Also, shouldn't the label for struct types include the tag of each field? Since differing tags make it a different struct type? Ideally this would have a test as well. This could be done as a follow-up, but it also fits in pretty naturally with this PR.

**Comment by smowton** [comment]
Note: at https://github.com/github/codeql/pull/17341 I added a stats update since there are new db tables and they ought to have associated stats. This evidently caused join-order problems since DCA flipped from just fine to catastrophic, so join order fixery (or simply accepting missing stats -- I note C# recently totally removed them and simply manually hacked their join orders where necessary) will be necessary before this can be merged.

**Comment by smowton** [comment]
@mbg @owen-mc all comments applied.

I have also taken the liberty of resurrecting https://github.com/github/codeql/pull/9386 and including it here since we have a dbscheme change afoot anyway.

File: go/ql/lib/semmle/go/Types.qll
Comment (owen-mc) [inline go/ql/lib/semmle/go/Types.qll:779]: It's confusing that you sometimes get the qualified name and sometimes the unqualified name. How about giving the qualified name in both cases? That would match the predicate name.

**Review by owen-mc — COMMENTED** [review]
A few optional suggestions and one change I feel more strongly about.

File: go/ql/lib/semmle/go/Types.qll
Comment (smowton) [reply go/ql/lib/semmle/go/Types.qll:785]: I've added a mention of `Exported` instead.

File: go/ql/lib/semmle/go/Types.qll
Comment (smowton) [reply go/ql/lib/semmle/go/Types.qll:779]: Not done: the goal here is to only apply qualifiers where needed to distinguish types, i.e. if the method is non-exported and declared in a different package.

Comment (owen-mc) [reply go/ql/lib/semmle/go/Types.qll:779]: Then maybe the predicate name should be changed from `getMethodTypeByQualifiedName`? At the moment it's just a recipe for confusion. Distinguishing name? Possibly qualified name?

Comment (mbg) [reply go/ql/lib/semmle/go/Types.qll:779]: I think I agree with @owen-mc here. It may be that the motivation for adding this was to distinguish types that otherwise appear equal, but that doesn't mean it has to be limited to the minimum necessary to accomplish that at the cost of having a confusingly-named predicate.

If the intention is to implement something just distinguish between different interfaces, I'd maybe call this something like `getIdentityMethodName`?

Also, the comment here could be clearer. There is some ambiguity in what "both" in "would yield ... for both" refers to. It's also not immediately obvious from the documentation how it relates to how the predicate is used.

Comment (smowton) [reply go/ql/lib/semmle/go/Types.qll]: I've simply deleted the predicate and moved the complexity to the user: https://github.com/github/codeql/pull/17357/commits/87c1e7decf83caffcee83771f12b1587f30b31f1

**Review by owen-mc — DISMISSED** [review]

**Comment by smowton** [comment]
(I'll hold off on merging this for now since there are still moderate performance issues brought about by the stats update)

**Comment by smowton** [comment]
DCA is now pretty good after the latest wave. Considering all the join-order tweaks needed based on the new stats, I'll do a QA run to get a larger performance sample.

**Comment by smowton** [comment]
QA results were broadly very strong -- analysis time reductions were much more common than increases -- but I've debugged a few of the notably somewhat-slower projects and made one last DCA run to retest the usual suite + those projects showing a slowdown on QA.

**Review by mbg — APPROVED** [review]
Happy with this for when you're happy with the performance results

**Review by owen-mc — APPROVED** [review]
Merge when performance is good enough.

**Review by owen-mc — APPROVED** [review]

**Review by owen-mc — APPROVED** [review]

**Review by owen-mc — APPROVED** [review]

**Comment by smowton** [comment]
At long last -- performance results are good. QA shows an overall -8.5% time spent running queries, though over half of that comes from one leviathan project whose 2h30 analysis now takes 15 minutes. There are a small number of QA projects that recurred across a few runs showing moderate-sized (20-second or so) increases in runtime, but which didn't have an obvious cause looking at predicate timing tables, or comparing the RA of the most expensive predicates against the RA generated on `main` -- my guess is that these are cases where scheduling and/or cache eviction is perturbed for the worse, and for now I'm going to have to let them go, to revisit if we see a more flagrant problem.

**Pull request merged**
